File: android-katha-ai/app/src/main/java/com/rork/kathaai/ui/screens/SupportComponents.kt
package com.rork.kathaai.ui.screens

import android.content.Intent
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Error
import androidx.compose.material.icons.outlined.PlayArrow
import androidx.compose.material.icons.outlined.KeyboardArrowUp
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.rork.kathaai.data.SeedData
import com.rork.kathaai.model.Story
import com.rork.kathaai.ui.components.TextLink
import com.rork.kathaai.ui.theme.KathaTheme
import com.rork.kathaai.viewmodel.AppViewModel
import com.rork.kathaai.viewmodel.KathaUiState
import com.rork.kathaai.viewmodel.openAudioPlayer

// MARK: - Audio Mini Bar (FIX 8)

@Composable
fun AudioMiniBar(
    story: Story,
    state: KathaUiState,
    viewModel: AppViewModel
) {
    val audioState = viewModel.audioState(story.id)
    val context = LocalContext.current

    when (audioState) {
        AppViewModel.AudioBarState.PREPARING -> PreparingBar(story, state, viewModel, context)
        AppViewModel.AudioBarState.READY -> ReadyBar(story, viewModel)
        AppViewModel.AudioBarState.ERROR -> ErrorBar(story, viewModel)
    }
}

@Composable
private fun PreparingBar(
    story: Story,
    state: KathaUiState,
    viewModel: AppViewModel,
    context: android.content.Context
) {
    val transition = rememberInfiniteTransition(label = "pulse")
    val pulseAlpha by transition.animateFloat(
        initialValue = 0.4f,
        targetValue = 1.0f,
        animationSpec = infiniteRepeatable(tween(1200), RepeatMode.Reverse),
        label = "pulseAlpha"
    )
    var hasTapped by remember { mutableStateOf(false) }

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(KathaTheme.surface)
            .clickable {
                if (!hasTapped) {
                    hasTapped = true
                    viewModel.showToast("Audio preparing — usually ready in 8–15 seconds")
                }
            }
            .padding(horizontal = KathaTheme.Spacing.l, vertical = KathaTheme.Spacing.m),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(KathaTheme.Spacing.m)
    ) {
        Box(
            modifier = Modifier
                .size(24.dp)
                .clip(CircleShape)
                .background(KathaTheme.accent.copy(alpha = pulseAlpha))
        )
        Column(modifier = Modifier.weight(1f)) {
            Text(
                story.title,
                color = KathaTheme.textPrimary,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                maxLines = 1
            )
            Text(
                "Preparing audio…",
                color = KathaTheme.textSecondary,
                fontSize = 12.sp
            )
        }
        Icon(
            Icons.Outlined.KeyboardArrowUp, null,
            tint = KathaTheme.textTertiary.copy(alpha = 0.4f),
            modifier = Modifier.size(20.dp)
        )
    }
}

@Composable
private fun ReadyBar(story: Story, viewModel: AppViewModel) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(KathaTheme.surface)
            .clickable { viewModel.openAudioPlayer(story.id) }
            .padding(horizontal = KathaTheme.Spacing.l, vertical = KathaTheme.Spacing.m),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(KathaTheme.Spacing.m)
    ) {
        Box(
            modifier = Modifier
                .size(32.dp)
                .clip(CircleShape)
                .background(KathaTheme.accent),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                Icons.Outlined.PlayArrow, "Play",
                tint = Color.White,
                modifier = Modifier.size(14.dp)
            )
        }
        Column(modifier = Modifier.weight(1f)) {
            Text(
                story.title,
                color = KathaTheme.textPrimary,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                maxLines = 1
            )
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(2.dp)
                    .clip(RoundedCornerShape(1.dp))
                    .background(KathaTheme.accent)
            ) {}
        }
        Icon(
            Icons.Outlined.KeyboardArrowUp, null,
            tint = KathaTheme.textSecondary,
            modifier = Modifier.size(20.dp)
        )
    }
}

@Composable
private fun ErrorBar(story: Story, viewModel: AppViewModel) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(KathaTheme.surface)
            .padding(horizontal = KathaTheme.Spacing.l, vertical = KathaTheme.Spacing.m),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(KathaTheme.Spacing.m)
    ) {
        Icon(
            Icons.Outlined.Error, null,
            tint = KathaTheme.error,
            modifier = Modifier.size(20.dp)
        )
        Column(modifier = Modifier.weight(1f)) {
            Text(
                story.title,
                color = KathaTheme.textPrimary,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                maxLines = 1
            )
            Text(
                "Audio unavailable",
                color = KathaTheme.error,
                fontSize = 12.sp
            )
        }
        TextLink(title = "Retry") {
            viewModel.retryAudio(story.id)
        }
    }
}

// MARK: - Share Feedback Helper (FIX 7)

// TODO: Replace with real feedback email address before launch
const val FEEDBACK_EMAIL = "feedback@katha.ai"

fun launchFeedbackEmail(
    context: android.content.Context,
    appVersion: String,
    username: String
) {
    val deviceModel = android.os.Build.MODEL
    val osVersion = android.os.Build.VERSION.RELEASE
    val body = """

---
App version: $appVersion
Platform: Android
Device: $deviceModel
OS: $osVersion
User: $username
"""
    val intent = Intent(Intent.ACTION_SENDTO).apply {
        data = android.net.Uri.parse("mailto:$FEEDBACK_EMAIL")
        putExtra(Intent.EXTRA_SUBJECT, "Katha AI Feedback — v$appVersion")
        putExtra(Intent.EXTRA_TEXT, body)
    }
    try {
        context.startActivity(Intent.createChooser(intent, "Share feedback"))
    } catch (e: Exception) {
        // Fallback handled by caller
    }
}

File: ios-katha-ai-create-stories/KathaAICreateStories/Views/SupportComponents.swift
//
//  SupportComponents.swift
//  KathaAICreateStories
//

import SwiftUI
import MessageUI
import UIKit

// MARK: - Feedback Mail Composer (FIX 7)

struct FeedbackMailComposer: UIViewControllerRepresentable {
    let appVersion: String
    let username: String

    // TODO: Replace with real feedback email address before launch
    static let feedbackEmail = "feedback@katha.ai"

    func makeUIViewController(context: Context) -> MFMailComposeViewController {
        let controller = MFMailComposeViewController()
        controller.mailComposeDelegate = context.coordinator
        controller.setToRecipients([Self.feedbackEmail])
        controller.setSubject("Katha AI Feedback — v\(appVersion)")
        let device = UIDevice.current.model
        let osVersion = UIDevice.current.systemVersion
        let body = """

---
App version: \(appVersion)
Platform: iOS
Device: \(device)
OS: \(osVersion)
User: \(username)
"""
        controller.setMessageBody(body, isHTML: false)
        return controller
    }

    func updateUIViewController(_ uiViewController: MFMailComposeViewController, context: Context) {}

    func makeCoordinator() -> Coordinator { Coordinator() }

    final class Coordinator: NSObject, MFMailComposeViewControllerDelegate {
        func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
            controller.dismiss(animated: true)
        }
    }
}

// MARK: - Audio Mini Bar (FIX 8)

enum AudioBarState {
    case preparing
    case ready
    case error
}

struct AudioMiniBar: View {
    @Environment(AppState.self) private var appState
    let story: Story

    @State private var pulseOpacity: Double = 0.4
    @State private var showReadyTransition: Bool = false

    private var audioState: AudioBarState {
        appState.audioState(for: story.id)
    }

    var body: some View {
        Group {
            switch audioState {
            case .preparing:
                preparingBar
            case .ready:
                readyBar
            case .error:
                errorBar
            }
        }
        .onAppear {
            if audioState == .preparing {
                startPulse()
                appState.startAudioPreparation(storyId: story.id)
            }
        }
    }

    // MARK: - Preparing

    private var preparingBar: some View {
        HStack(spacing: KathaTheme.Spacing.m) {
            Circle()
                .fill(KathaTheme.accent)
                .frame(width: 24, height: 24)
                .opacity(pulseOpacity)

            VStack(alignment: .leading, spacing: 2) {
                Text(story.title)
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(KathaTheme.textPrimary)
                    .lineLimit(1)
                Text("Preparing audio…")
                    .font(.system(size: 12))
                    .foregroundStyle(KathaTheme.textSecondary)
            }

            Spacer()

            Image(systemName: "chevron.up")
                .font(.system(size: 16))
                .foregroundStyle(KathaTheme.textTertiary.opacity(0.4))
        }
        .padding(.horizontal, KathaTheme.Spacing.l)
        .padding(.vertical, KathaTheme.Spacing.m)
        .background(KathaTheme.surface)
        .contentShape(Rectangle())
        .onTapGesture {
            Haptics.light()
            appState.showToast("Audio preparing — usually ready in 8–15 seconds")
        }
    }

    // MARK: - Ready

    private var readyBar: some View {
        HStack(spacing: KathaTheme.Spacing.m) {
            Button {
                Haptics.light()
                appState.showToast("Audio playback coming in the next update ✨")
            } label: {
                ZStack {
                    Circle()
                        .fill(KathaTheme.accent)
                        .frame(width: 32, height: 32)
                    Image(systemName: "play.fill")
                        .font(.system(size: 14))
                        .foregroundStyle(.white)
                }
            }

            VStack(alignment: .leading, spacing: 2) {
                Text(story.title)
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(KathaTheme.textPrimary)
                    .lineLimit(1)
                RoundedRectangle(cornerRadius: 1)
                    .fill(KathaTheme.accent)
                    .frame(height: 2)
                    .frame(maxWidth: .infinity)
            }

            Image(systemName: "chevron.up")
                .font(.system(size: 16))
                .foregroundStyle(KathaTheme.textSecondary)
        }
        .padding(.horizontal, KathaTheme.Spacing.l)
        .padding(.vertical, KathaTheme.Spacing.m)
        .background(KathaTheme.surface)
        .transition(.opacity.animation(.easeInOut(duration: 0.3)))
    }

    // MARK: - Error

    private var errorBar: some View {
        HStack(spacing: KathaTheme.Spacing.m) {
            Image(systemName: "exclamationmark.circle")
                .font(.system(size: 20))
                .foregroundStyle(KathaTheme.error)

            VStack(alignment: .leading, spacing: 2) {
                Text(story.title)
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(KathaTheme.textPrimary)
                    .lineLimit(1)
                Text("Audio unavailable")
                    .font(.system(size: 12))
                    .foregroundStyle(KathaTheme.error)
            }

            Spacer()

            TextLink(title: "Retry") {
                appState.retryAudio(storyId: story.id)
            }
        }
        .padding(.horizontal, KathaTheme.Spacing.l)
        .padding(.vertical, KathaTheme.Spacing.m)
        .background(KathaTheme.surface)
    }

    // MARK: - Pulse Animation

    private func startPulse() {
        withAnimation(.easeInOut(duration: 1.2).repeatForever(autoreverses: true)) {
            pulseOpacity = 1.0
        }
    }
}
